Bug 1109806 - send links to desktop via Handoff

TETRA2000 opened the pull request.

st3fan commented on Client/Frontend/Browser/BrowserViewController.swift:
```diff
@@ -98,6 +99,7 @@ class BrowserViewController: UIViewController {
         self.profile = profile
         self.tabManager = tabManager
         self.readerModeCache = DiskReaderModeCache.sharedInstance
+        self.handoffManager = HandoffManager.sharedInstance
```
If there is a shared instance of HandoffManager, we don't need to have a reference to it permanently in BrowserViewController. I would just replace all occurrences of `handoffManger` to `HandoffManager.sharedInstance`.

fluffyemily commented on Client/Frontend/Browser/HandoffManager.swift:
```diff
@@ -0,0 +1,45 @@
+* file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+class HandoffManager: NSObject {
+    private static var _sharedInstance: HandoffManager?
```
There is no need to have a private var and a public one, you can just create the singleton instance with

`static var sharedInstance = HandoffManager()`

fluffyemily commented on Client/Frontend/Browser/HandoffManager.swift:
```diff
@@ -0,0 +1,45 @@
+    func stop() {
+        if #available(iOS 9.0, *) {
+            userActivity?.resignCurrent()
+        } else {
```
nit: I would prefer not to have an empty else statement here. Either add a log statement of remove the else and leave a comment above the if statement

fluffyemily commented on Client/Frontend/Browser/HandoffManager.swift:
```diff
@@ -0,0 +1,45 @@
+    lazy var userActivity: NSUserActivity? = NSUserActivity(activityType: NSUserActivityTypeBrowsingWeb)
+    
+    func start() {
+        if let _ = userActivity?.webpageURL {
```
nit: would be nicer as a `guard` rather than an `if let`

TETRA2000 commented:
@st3fan @fluffyemily Thank you very much for your review!!
I fixed the patch. And also fixed the previous commit message.
After you confirm this, I'll squash these two commits.

sleroux commented on Client/Frontend/Browser/HandoffManager.swift:
```diff
@@ -0,0 +1,36 @@
+class HandoffManager: NSObject {
+    static var sharedInstance = HandoffManager()
+    
+    lazy var userActivity: NSUserActivity? = NSUserActivity(activityType: NSUserActivityTypeBrowsingWeb)
```
When using a lazy var you actually don't need to declare it as an optional if you wrap the result in a block. For example you can change this to:

```swift
lazy var userActivity: NSUserActivity = {
    return NSUserActivity(activityType: NSUserActivityTypeBrowsingWeb)
}()
```
Now you don't need to sprinkle questions marks throughout!

sleroux commented on Client/Frontend/Browser/HandoffManager.swift:
```diff
@@ -0,0 +1,36 @@
+        guard let urlStr = urlStr,
+            let url = NSURL(string: urlStr) else {
+                return
+        }
```
You can simplify this down since we have this neat extension method on Strings for converting into a URL:

```swift
guard let url = urlStr?.asURL else {
    return
}
```

sleroux commented on Client/Frontend/Browser/HandoffManager.swift:
```diff
@@ -0,0 +1,36 @@
+    
+    func stop() {
+         // iOS 8.x is not current target
+        if #available(iOS 9.0, *) {
```
How come we check the API level for stop but not on start?

sleroux commented on Client/Frontend/Browser/BrowserViewController.swift:
```diff
@@ -523,6 +528,9 @@ class BrowserViewController: UIViewController {
         screenshotHelper.viewIsVisible = false
+        
+        log.debug("BVC stopping handoff.")
+        HandoffManager.sharedInstance.stop()
```
I noticed that in all of the other calls to stop we also clear the URL but not here. How come? Would it make sense to clear every time we stop?

TETRA2000 replied:
@sleroux

I considered this operation.

Steps
* 1. View a web page. (not blank tab)
* 2. Press the tab switch button. (on the right top)
  - 2-a. Press the config button.
  - 2-b. Press the Incognito Mode button.


I think handoff should stop in the following situation. 
This code is intended to stop this.

2.
User may not understand which page in the tab list was sent
2-a.
Sending a link while showing the config page isn't intended.
2-b.
Users may misunderstand that link of incognito page was sent.


---
2
![2](https://cloud.githubusercontent.com/assets/1459450/12699902/39e1d116-c811-11e5-9e2e-b72f6ecfaf48.png)


2a
![2a](https://cloud.githubusercontent.com/assets/1459450/12699906/4590cd14-c811-11e5-9378-480e2e637c3d.png)

2b
![2b](https://cloud.githubusercontent.com/assets/1459450/12699907/45917188-c811-11e5-8608-dd7cedab79f9.png)

TETRA2000 commented:
@sleroux
Thanks to your review, codes in this patch became far more beautiful than before!!

* I fixed this patch according to your review.
* I also changed the `HandoffManager.swift`'s position in alphabet order.

_diff after your review_
https://github.com/TETRA2000/firefox-ios/commit/2063883b1f74ecc0ab9c9efa4bcc291d82a122c4

Thank you!!

sleroux commented:
Hey Takahiko,

Thanks for updating the patch! I've been doing some additional testing on this and found a couple of things:

I've been able to run into a case where I navigated to an about: url which resulted in the HandoffManager throwing an exception inside updateCurrentURL:
```
2016-02-01 10:17:05.791 Client[3533:1228278] *** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: 'NSUserActivity.webpageURL scheme "about" is not allowed.'
*** First throw call stack:
(0x182fd1900 0x18263ff80 0x19248ec58 0x1924887ac 0x1001b65f4 0x100326e70 0x100326070 0x1003256d0 0x100326024 0x1838bc804 0x1838bc32c 0x1838a6b44 0x18a82c4e4 0x18a9285ac 0x18a946f34 0x18a7f6fd4 0x18a97cf8c 0x18a7b214c 0x18a7b41f8 0x186793218 0x1867936c4 0x182f88efc 0x182f88990 0x182f86690 0x182eb5680 0x1843c4088 0x187d2cd90 0x10030dd08 0x182a568b8)
libc++abi.dylib: terminating with uncaught exception of type NSException
```
We don't support about: urls yet but I think we need to provide some sort of validation on the URL before we pass it to NSUserActivity. For now it would be safe to only allow http/https scheme urls.

I also noticed that in the scenarios you described above regarding stopping Handoff when in the Tab tray that Handoff still appears. I ran the build on device, opened a new tab and went back into the tab tray and was still able to open using Handoff on desktop.

Lastly, after opening the tab using Handoff on desktop the app becomes unresponsive. I have a feeling that these issues might only be occurring on device and not on the simulator. Are you able to run the app in the device? It can be challenging at the moment with how our configuration works but I can assist in getting that to work if you need help debugging.

TETRA2000 commented:
@sleroux 
Thank you very much for your testing!!

I agree with adding validation for http/https.
Soon I’ll update my patch.

Thanks for pointing out these problems.
They will really helpful for me to improve the patch.
I’ll confirm and fix soon.

I've already succeed to ran this app on my device.
(Screenshots of my test scenario are also taken with a real device. I used QuickTime.)

TETRA2000 commented:
@sleroux 

I updated my patch to validate passed URL.

> I also noticed that in the scenarios you described above regarding stopping Handoff when in the Tab tray that Handoff still appears. I ran the build on device, opened a new tab and went back into the tab tray and was still able to open using Handoff on desktop.

I think this is a timing issue.
After returned to tab menu, it will take a few seconds to disappear app icon in task tray.

> Lastly, after opening the tab using Handoff on desktop the app becomes unresponsive. I have a feeling that these issues might only be occurring on device and not on the simulator. Are you able to run the app in the device? It can be challenging at the moment with how our configuration works but I can assist in getting that to work if you need help debugging.

I tried to reproduce this issue but I couldn't.
Please give me more information.
Desktop browser name, iPhone model type, OS version, URLs, etc...

Thank you!!

TETRA2000 commented:
I rebased my patch with latest master.

sleroux commented:
Thanks for rebasing! Unfortunately we made some large project changes last week that will require some additional rebasing. Let me me know if you run into issues with that.

I ran this branch again and tried to open the 'What's New' tab that we launch on app updates using Handoff and experienced an assertion error here:

![screenshot 2016-02-17 10 28 30](https://cloud.githubusercontent.com/assets/434322/13114451/86370bf0-d561-11e5-9678-807d96234961.png)

What looks like is happening is that the new work we've done for handling Spotlight/3D Touch handling conflicts with the Handoff code since they both use NSUserActivity. The assumption it seems we make is that if an activity is being continued from here [1] we create a new tab which we don't want to happen in the case of Handoff. At the moment, handling NSUserActivities in the app is pretty complicated after implementing 3D Touch/Spotlight for the upcoming release. I think we'll need to figure out how to coordinate adding Handoff along with refactoring how we handle Spotlight/3D Touch to have a unified way of working with NSUserActivities.

@fluffyemily @jhugman Any thoughts on how we would incorporate Handoff with the existing NSUserActivity work we did?

[1] https://github.com/mozilla/firefox-ios/blob/master/Client/Helpers/SpotlightHelper.swift#L115

TETRA2000 commented:
@sleroux Thank you for your information!!
While I'm checking new changes of Spotlight/3D Touch, I found that these changes already enabled Handoff. But there is some problems like, 
* Clicking the Handoff icon after I opened preference page in my device, pref page crashes.
* Clicking the Handoff icon while using incognito tab, tab crashes.(And sends wrong link.)
* The Handoff icon still appears when I opened blank tab.

I also confirmed these problems on release version.

I'm finding solution for these issues.

sleroux commented:
Hey @TETRA2000!

Figured I would drop in and see how things are going. Looks like it's been pretty quiet for the pas few months. Any interest in picking this up further? If not, I'll free up the bug for another contributor and close the PR for later reference.

farhanpatel commented:
Thanks for this PR! We just finished a major migration to swift 3.0 making merging in most patches pretty difficult. If you feel like this patch is still important please feel free to update your PR to swift 3 and reopen this PR 😄
